clinkc/sample/upnpavdump/macosx/xcode30/upnpavdump.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 08/05/20               \" DATE 
.Dt upnpavdump 1      \" Program name and manual section number 
.Os Darwin
.Sh NAME                 \" Section Header - required - don't modify 
.Nm upnpavdump,
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line. 
.Nm Other_name_for_same_program(),
.Nm Yet another name for the same program.
.\" Use .Nm macro to designate other names for the documented program.
.Nd This line parsed for whatis database.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl abcd              \" [-abcd]
.Op Fl a Ar path         \" [-a path] 
.Op Ar file              \" [file]
.Op Ar                   \" [file ...]
.Ar arg0                 \" Underlined argument - use .Ar anywhere to underline
arg2 ...                 \" Arguments
.Sh DESCRIPTION          \" Section Header - required - don't modify
Use the .Nm macro to refer to your program throughout the man page like such:
.Nm
Underlining is accomplished with the .Ar macro like this:
.Ar underlined text .
.Pp                      \" Inserts a space
A list of items with descriptions:
.Bl -tag -width -indent  \" Begins a tagged list 
.It item a               \" Each item preceded by .It macro
Description of item a
.It item b
Description of item b
.El                      \" Ends the list
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl a                 \"-a flag as a list item
Description of -a flag
.It Fl b
Description of -b flag
.El                      \" Ends the list
.Pp
.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "/Users/joeuser/Library/really_long_file_name" -compact
.It Pa /usr/share/file_name
FILE_1 description
.It Pa /Users/joeuser/Library/really_long_file_name
FILE_2 description
.El                      \" Ends the list
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr a 1 , 
.Xr b 1 ,
.Xr c 1 ,
.Xr a 2 ,
.Xr b 2 ,
.Xr a 3 ,
.Xr b 3 
.\" .Sh BUGS              \" Document known, unremedied bugs 
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner

clinkc/ChangeLog
2011-09-16 Satoshi Konno <skonno@cybergarage.org>
	* 2.4

2011-09-10 Satoshi Konno <skonno@cybergarage.org>
	* Fixed to compile using XCode4 on MacOSX and iOS normally.
	* Added source codes for Objective-C
	* Removed TARGET_OS_IPHONE define from the compiler option setting.

2011-03-09 Satoshi Konno <skonno@cybergarage.org>
	* Imporved cg_log_print() to return immediately when the output target is nothing.
	* Added cg_upnpav_content_setupnpalbumarturi() and cg_upnpav_content_getupnpalbumarturi().

2011-02-23 Satoshi Konno <skonno@cybergarage.org>
	* Added the control point parameter into CG_UPNP_DEVICE_LISTENER.
	* Added cg_upnp_ssdprequest_print() and cg_upnp_ssdpresponse_print().
	* Added some debug messages.
	* Added CGUpnpDeviceDelegate (Objective-C).
	
2010-10-26  Satoshi Konno <skonno@cybergarage.org>
	* Added cg_upnp_controlpoint_isrunning().

2010-08-20  Satoshi Konno <skonno@cybergarage.org>
	* Changed some source codes to compile normally using Visual Studio 2008.
	* Added some project files for Visual Studio 2008.
	* Added cg_upnpav_dms_updatenetworkinterfaces() and cg_upnpav_dms_getnetworkinterface();

2010-08-17  Ariel Yaloz <arielyz@hotmail.com>
	* Added support for getting the list of allowable values for a state variable
	* Added a method for checking whether a state variable is implemented
	* Fix compilation on MacOS X
	
2010-07-31 Roman Shterenzon <romanbsd@yahoo.com>
	* Fixed autoconf warnings
	* Don't append the original query to path when it already has a query (fixes WMP11/12 access)
	* Fixed several memory leaks in Objective-C wrapper
	* Proper handling of edge cases in src/cybergarage/upnp/cservice.c and
	  std/av/src/cybergarage/upnp/std/av/cprotocolinfo.c
	* Fix crashes caused by attempts to initialize NSString with NULL cstring
	* Added modelNumber and modelName methods to CGUpnpDevice class

2010-06-28 Christian Zuckschwerdt <Christian@Zuckschwerdt.org>
	* Fixed CGUpnpAction::arguments() not to release temporary object.
	
2010-06-24  Satoshi Konno <skonno@cybergarage.org>
	* Added cg_xml_node_print().

2010-03-17  Satoshi Konno <skonno@cybergarage.org>
	* Added cg_upnp_service_createnotifyallthread().
	* Changed cg_upnp_device_newsubscriptionrecieved() to delay the notify message instead of cg_upnp_service_notifyall().
	* Changed cg_upnp_event_notify_request_setpropertysetnode() to add a connection header.
	* Changed cg_http_getservername() to use uname() on iPhone platform.

2010-02-09  Satoshi Konno <skonno@cybergarage.org>
	* Added project files for Visual Studio 2008.
	* Disabled AC_PROG_LIBTOOL and HAVE_LIBTOOL in confugre.in and Makefile.am.
	* Disabled libxml2 and libcurl as default in confugre.in.
	* Added support for SO_NOSIGPIPES in configure.in.
	* Fixed some errors to compile normally on Ubuntu 9.10.
	* Changed CG_USE_WIN32_GETHOSTADDRESSES flag as default for Windows platform.
	* Fixed cg_upnp_device_unsubscriptionrecieved() to unscribe normally.
	* Fixed cg_upnp_device_ssdpmessagereceived() not to ignore IPv6 multicast.
	* Updated cg_upnp_createuuid() to use naitive functions on Windows and Unix platforms.
	
2010-01-20  Satoshi Konno <skonno@cybergarage.org>
	* Removed service mutex control from cg_upnp_service_addsubscriber() to avoid deadlock.

2009-09-01  Satoshi Konno <skonno@cybergarage.org>
	* Added cg_soap_response_getstatuscode().
	* Added cg_upnp_event_notify_response_getstatuscode().

2009-08-27  Satoshi Konno <skonno@cybergarage.org>
	* Added cg_upnp_argument_setfloatvalue() and cg_upnp_argument_getfloatvalue().
	* Added cg_upnp_argument_setdoublevalue() and cg_upnp_argument_getdoublevalue().
	* Added cg_string_getfloatvalue() and cg_string_getdoublevalue().
	* Added cg_float2str() and cg_double2str().
	* Added cg_str2float() and cg_str2double().
	* Added cg_upnp_device_gethttpserverlist() and cg_upnp_device_getssdpserverlist().
	* Added cg_http_server_getsocket().
	* Added CGUpnpDevice::isRunning().

2009-06-01  Satoshi Konno <skonno@cybergarage.org>
	* Added cg_dictionary_*();
	* Added cg_net_uri_getquerydictionary().
	* Fixed a buffer overrun of cg_xml_escapechars().
	* Changed cg_socket_setid() not to raise SIGPIPE on iPhone.
	
2009-05-28  Satoshi Konno <skonno@cybergarage.org>
	* Improved cg_string_*() and cg_list_*() not to abort when the specified string object is NULL.
	* Added cg_upnp_device_getsmallesticonbymimetype().
	* Added cg_upnp_device_addicon().
	* Added CGUpnpDevice:setDeviceType().
	* Added cg_http_request_postdata().
	* Added cg_http_request_postchunkeddata() and cg_http_request_postlastchunk().
	* Added cg_http_request_closesocket().
	* Added cg_upnp_device_isrunning().

2009-04-23  Satoshi Konno <skonno@cybergarage.org>
	* Added DLNA-CP-version in the USER-AGENT HTTP header value as default.

2009-01-29  Satoshi Konno <skonno@cybergarage.org>
	* Added cg_net_uri_rebuild() and cg_net_url_rebuild().
	* Fixed cg_upnp_service_mangleurl() to return correct url using cg_net_uri_rebuild() when a device has URLBase and the service's URL is relative.
	* Changed cg_upnp_controlpoint_parsescservicescpd() to read the service description internally because the some illegal devices, such as MediaLink, don't return the specified service description.

2008-12-08  Satoshi Konno <skonno@cybergarage.org>
	* Added cg_upnp_device_getsmallesticon().
	* Added cg_upnp_device_getabsoluteiconurl().
	* Added cg_net_uri_setvalue() and cg_net_uri_getvalue().
	* Deprecated cg_net_uri_set().
	* Added CGUpnpDevice:urlBase().
	* Added CGUpnpDevice:smallestIcon().
	* Added CGUpnpDevice:absoluteIconUrl().
	
2008-11-18  Satoshi Konno <skonno@cybergarage.org>
	* Changed cg_http_request_post_main() to write the first line at a time because SIGPIPE occurred on some HTTP server such as Flickr.

2008-10-31  Satoshi Konno <skonno@cybergarage.org>
	* v2.3
	* Added UPnP/AV package in std/av as a sample implementation.

2008-09-26 Satoshi Konno <skonno@cybergarage.org>
	* Fixed the following methods correctly

clinkc/std/av/sample/upnpavdump/upnpavdump.c
/************************************************************
*
*	CyberLink for C
*
*	Copyright (C) Satoshi Konno 2008
*
*	File: upnpavdump.c
*
*	Revision:
*       08/05/24
*               - first release.
*
************************************************************/

#include <cybergarage/upnp/cupnp.h>

#define UPNPAVDUMP_DMS_DEVICETYPE "urn:schemas-upnp-org:device:MediaServer:1"
#define UPNPAVDUMP_DMS_CONTENTDIR_SERVICETYPE "urn:schemas-upnp-org:service:ContentDirectory:1"
#define UPNPAVDUMP_DMS_BROWSE_ACTIONNAME "Browse"

/////////////////////////////////////////////////////////////////////////////////
// PrintContentDirectory
/////////////////////////////////////////////////////////////////////////////////

void PrintContentDirectory(CgUpnpAction *browseAction, int indent, char *objectId)
{
	int n;
	char indentStr[128];
	char *resultXml;
	CgXmlParser *xmlParser;
	CgXmlNodeList *rootNode;
	CgXmlNode *didlNode;
	CgXmlNode *cnode;
	char *id;
	char *title;
	char *url;

	for (n=0; n<indent && n<(sizeof(indentStr)-1); n++)
		indentStr[n] = ' ';
	indentStr[n] = '\0';

	cg_upnp_action_setargumentvaluebyname(browseAction, "ObjectID", objectId);
	cg_upnp_action_setargumentvaluebyname(browseAction, "BrowseFlag", "BrowseDirectChildren");
	cg_upnp_action_setargumentvaluebyname(browseAction, "Filter", "*");
	cg_upnp_action_setargumentvaluebyname(browseAction, "StartingIndex", "0");
	cg_upnp_action_setargumentvaluebyname(browseAction, "RequestedCount", "0");
	cg_upnp_action_setargumentvaluebyname(browseAction, "SortCriteria", "");
	
	if (!cg_upnp_action_post(browseAction))
		return;

	resultXml = cg_upnp_action_getargumentvaluebyname(browseAction, "Result");
	if (cg_strlen(resultXml) <= 0)
		return;

	rootNode = cg_xml_nodelist_new();
	xmlParser = cg_xml_parser_new();
	if (cg_xml_parse(xmlParser, rootNode, resultXml, cg_strlen(resultXml))) {
		didlNode = cg_xml_nodelist_getbyname(rootNode, "DIDL-Lite");
		if (didlNode) {
			for (cnode=cg_xml_node_getchildnodes(didlNode); cnode; cnode=cg_xml_node_next(cnode)) {
				id = cg_xml_node_getattributevalue(cnode, "id");
				title = cg_xml_node_getchildnodevalue(cnode, "dc:title");
				if (cg_xml_node_isname(cnode, "container")) {
					printf(" %s[%s]%s\n", 
						indentStr, 
						id, 
						((0 < cg_strlen(title)) ? title : ""));
					PrintContentDirectory(browseAction, (indent+1), id);
				}
				else {
					url = cg_xml_node_getchildnodevalue(cnode, "res");
					printf(" %s[%s]%s (%s)\n", 
						indentStr, 
						id, 
						((0 < cg_strlen(title)) ? title : ""),
						((0 < cg_strlen(url)) ? url: ""));
				}
			}
		}
	}
	cg_xml_nodelist_delete(rootNode);
	cg_xml_parser_delete(xmlParser);
}

/////////////////////////////////////////////////////////////////////////////////
// PrintDMSInfo
/////////////////////////////////////////////////////////////////////////////////

void PrintDMSInfo(CgUpnpDevice *dev, int dmsNum)
{
	CgUpnpService *conDirService;
	CgUpnpAction *browseAction;
	CgUpnpStateVariable *searchCap;
	CgUpnpStateVariable *sorpCap;

	if (!cg_upnp_device_isdevicetype(dev, UPNPAVDUMP_DMS_DEVICETYPE))
		return;
	
	printf("[%d] : %s\n", dmsNum, cg_upnp_device_getfriendlyname(dev));

	conDirService = cg_upnp_device_getservicebytype(dev, UPNPAVDUMP_DMS_CONTENTDIR_SERVICETYPE);
	if (!conDirService)
		return;

	searchCap = cg_upnp_service_getstatevariablebyname(conDirService, "SearchCapabilities");
	if (searchCap) {
		if (cg_upnp_statevariable_post(searchCap))
			printf(" SearchCapabilities = %s\n", cg_upnp_statevariable_getvalue(searchCap));
	}

	sorpCap = cg_upnp_service_getstatevariablebyname(conDirService, "SortCapabilities");
	if (sorpCap) {
		if (cg_upnp_statevariable_post(sorpCap))
			printf(" SortCapabilities = %s\n", cg_upnp_statevariable_getvalue(sorpCap));
	}

	browseAction = cg_upnp_service_getactionbyname(conDirService, UPNPAVDUMP_DMS_BROWSE_ACTIONNAME);
	if (!browseAction)
		return;

	PrintContentDirectory(browseAction, 0, "0");
}

/////////////////////////////////////////////////////////////////////////////////
// PrintDMSInfos
/////////////////////////////////////////////////////////////////////////////////

void PrintDMSInfos(CgUpnpControlPoint *ctrlPoint)
{
	CgUpnpDevice *dev;
	int dmsNum;
		
	dmsNum = 0;
	for (dev = cg_upnp_controlpoint_getdevices(ctrlPoint); dev != NULL; dev = cg_upnp_device_next(dev)) {
		if (cg_upnp_device_isdevicetype(dev, UPNPAVDUMP_DMS_DEVICETYPE))
			PrintDMSInfo(dev, ++dmsNum);
	}

	if (dmsNum <= 0)
		printf("Media Server is not found !!\n");
}

/////////////////////////////////////////////////////////////////////////////////
// main
/////////////////////////////////////////////////////////////////////////////////

#if defined(TENGINE)
MBEG
#else
int main( int argc, char* argv[] )
#endif
{
	CgUpnpControlPoint *ctrlPoint;

	ctrlPoint = cg_upnp_controlpoint_new();
	if (cg_upnp_controlpoint_start(ctrlPoint) == FALSE) {
		printf("Couldn't start this control point !!");
		exit(1);
	}
	
	cg_upnp_controlpoint_search(ctrlPoint, CG_UPNP_ST_ROOT_DEVICE);

	cg_sleep(cg_upnp_controlpoint_getssdpsearchmx(ctrlPoint) * 1000);

	PrintDMSInfos(ctrlPoint);
	
	cg_upnp_controlpoint_stop(ctrlPoint);
	cg_upnp_controlpoint_delete(ctrlPoint);
	
	return(0);
}
